be/src/vec/exec/join/vhash_join_node.cpp

```diff
@@ -168,46 +177,98 @@ struct ProcessHashTableProbe {
             }
-
             int repeat_count = 0;
-            auto find_result =
-                    (*null_map)[_probe_index]
+            if constexpr (is_inner_join) {
+                if (!(*null_map)[_probe_index]) {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+
+                    if (find_result.is_found()) {
+                        auto& mapped = find_result.get_mapped();
+
+                        // TODO: Iterators are currently considered to be a heavy operation and have a certain impact on performance.
+                        // We should rethink whether to use this iterator mode in the future. Now just opt the one row case
+                        if (mapped.get_row_count() == 1) {
+                            mapped.visited = true;
+                            // right semi/anti join should dispose the data in hash table
+                            // after probe data eof
+                            ++repeat_count;
+                            for (size_t j = 0; j < _right_col_len; ++j) {
+                                auto& column = *mapped.block->get_by_position(j).column;
+                                mcol[j + _right_col_idx]->insert_from(column, mapped.row_num);
+                            }
+                        } else {
+                            if (_probe_index + 2 < _probe_rows)
+                                key_getter.prefetch(hash_table_ctx.hash_table, _probe_index + 2, _arena);
+                            for (auto it = mapped.begin(); it.ok(); ++it) {
+                                // right semi/anti join should dispose the data in hash table
+                                // after probe data eof
+                                ++repeat_count;
+                                for (size_t j = 0; j < _right_col_len; ++j) {
+                                    auto& column = *it->block->get_by_position(j).column;
+                                    // TODO: interface insert from cause serious performance problems
+                                    //  when column is nullable. Try to make more effective way
+                                    mcol[j + _right_col_idx]->insert_from(column, it->row_num);
+                                }
+                                it->visited = true;
+                            }
+                        }
+                    }
+                }
+            } else if constexpr (is_left_anti_join) {
+                if ((*null_map)[_probe_index]) { 
+                    ++repeat_count;
+                } else {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+                    if (find_result.is_found()) { 
+                        //do nothing
+                    } else {
+                        ++repeat_count;
+                    }
+                }
+            } else if constexpr(is_left_semi_join) {
```

Comment:
why not contexpr more join type, like right semi/anti join ?

Reply:
left semi , left anti, inner join is mostly used.so i just eliminate the three branch. others join will benefit from it